### DISCUSSION TITLE
Slow repeatable jobs without overlapping among multiple workers

### DISCUSSION THREAD

**Discussion by andrew-sol** [taskforcesh/bullmq#1409] [Q&A] [answered]
1. I have a repeatable job that runs every 30 seconds
2. I have many workers in different processes

The job might take longer than 30 seconds to execute, and the problem here is that another worker picks up the same job while the previous one has not finished yet. This might create race conditions in my DB.

I read [this](https://docs.bullmq.io/guide/jobs/repeatable#slow-repeatable-jobs) and I understand that it's expected behavior.

Is there a way to make all workers wait until the job finishes processing and let them start the job again only after that?

**Comment by manast** [comment]
There are two solutions I can think about to solve this requirement. 1) create a field in the DB to signal that a job is currently processing, and just return doing nothing in that case, 2) Use BullMQ Pro groups functionality with one group and max concurrency 1: https://docs.bullmq.io/bullmq-pro/groups/concurrency

**Comment by andrew-sol** [comment, marked as answer]
Thanks for the suggestions.

Currently, I implemented another solution. I always use constant `jobId` for all my repeatable jobs. I added a check into my worker that fetches all active jobs from the queue and checks whether there is another instance with the same `jobId` currently running; if there is, it skips the execution of the current one.

Here it is:
```ts
async canHandleJob(job: Job): Promise<boolean> {
  if (!job.opts.repeat) {
    return true;
  }

  const queue = this.queueService.getQueue(job.queueName);
  const activeJobs = await queue.getActive();

  for (const activeJob of activeJobs) {
    if (
      activeJob.id !== job.id &&
      activeJob.opts.repeat &&
      activeJob.opts.repeat.jobId &&
      activeJob.opts.repeat.jobId === job.opts.repeat.jobId
    ) {
      return false;
    }
  }

  return true;
}
```

It seems to be working fine. Or may it have any downsides?

**Reply by manast** [reply]
LGTM.
